[MRG] Refactor model_selection._search to include transductive estimators

jamiebull1 opened the pull request:
This generalises `BaseSearchCV` as `BaseSearch` so that it can be used on unsupervised learning where cross-validation does not make sense. The approach is to use a pair of mixins, `SearchClusterMixin` and `SearchCVMixin` to add `_fit` methods appropriate to each type of search.

The existing `GridSearchCV` and `RandomizedSearchCV` now inherit `BaseSearch` and `SearchCVMixin` while the new `GridSearchCluster` and `RandomizedSearchCluster` inherit `BaseSearch` and `SearchClusterMixin`.

See #6154 for a little more background.

jamiebull1 committed:
Refactor model_selection._search to include unsupervised estimators

jamiebull1 committed:
Refactor model_selection._search to include unsupervised estimators

jamiebull1 commented:
This is my first PR here so if someone could let me know how to find out what Travis and AppVeyor are failing on that would be really helpful.

jnothman commented:
It looks like not everything you want was committed here... I presume that's at least part of why you closed. Please feel free to reopen and ask for assistance with git if that's the main issue.

I look forward to reviewing this, but I'm very likely to push for something that's general to transductive algorithms rather than clusterers alone (e.g. including t-SNE).

jamiebull1 committed:
Skip GridSearchCluster and RandomizedSearchCluster in estimator_checks

jamiebull1 committed:
Skip GridSearchCluster and RandomizedSearchCluster in estimator_checks

jnothman commented:
It looks like the main changes you're hoping to contribute aren't in the patch yet. If this is something that's not ready yet for review, please prefix the issue title by `[WIP]`.

jamiebull1 commented:
What's missing? The main changes are in refactoring `model_selection._search` and adding the two new classes.

The reason I closed the PR is because I'd neglected to run the tests outside the `model_selection` package so missed that there were a few failing tests where the `GridSearchCluster` and `RandomizedSearchCluster` were getting picked up as regular estimators in the `tests.test_common` module. I assume the right thing to do here is to add them to the `OTHER` estimators in `utils.testing`? And also add them the skipped names in `estimator_checks.check_get_params_invariance`?

That's what I've done in any case, but I'm happy to discuss generalising further.

jnothman commented:
Ah right, it's not showing your changes in github because you've changed the files to use DOS line endings. Hence its diffs conclude that you've changed every line in the file. Please remove CR characters, commit and push.

jamiebull1 committed:
Generalise to non-clustering transductive estimators

Renames GridSearchCluster, etc to GridSearchTransductive, etc.
Adds a test for a non-clustering transductive model (t-SNE)

jamiebull1 committed:
Generalise to non-clustering transductive estimators

Renames GridSearchCluster, etc to GridSearchTransductive, etc.
Adds a test for a non-clustering transductive model (t-SNE)

jamiebull1 commented:
I hadn't seen t-SNE before, but I see that the fitted model can be scored on at least `kl_divergence_` and `trustworthiness`, and potentially on other custom metric. Given that these can quite easily be wrapped in scorers with the format `scorer(estimator, X, y=None)` it  works fine as-is with the new classes. This should work for other transductive algorithms too, assuming they can all be scored in some way.

So what I've done is:

1. Changed the names for `GridSearchCluster`, `RandomizedSearchCluster`, and `ClusterSearchMixin` to indicate that they can be used for any transductive model which can be scored. These are now `GridSearchTransductive`, etc.

2. Written a `trustworthiness_scorer` for `t-SNE` in the tests and used it to test a non-clustering transductive model in `test_search.test_grid_search_transductive_no_score` which is now passing.

jamiebull1 commented:
Sorry, just seen your comment. I'll change and push again.

jnothman commented:
Sounds great! I await a readable diff :)

jamiebull1 committed:
Remove DOS line endings

jamiebull1 committed:
Remove DOS line endings

jamiebull1 committed:
Fix _Score_Tuple.__repr__

This was still the one from the CV version.

jamiebull1 committed:
Fix _Score_Tuple.__repr__

This was still the one from the CV version.

jamiebull1 committed:
Replace examples for GridSearchCV deleted in error

jamiebull1 committed:
Replace examples for GridSearchCV deleted in error

jnothman commented:
Ta.

jamiebull1 commented:
Thanks for taking a look at this. I've just spotted a couple of errors now I can see the diff - the _ScoreTuple.__repr__ and the accidentally deleted examples for the original GridSearchCV. I'll fix them once I've got your comments back as I'm hitting the hay now. Looking forward to your comments.

jnothman commented:
Sleep well, and thanks. I don't promise substantial comments will come
today. Perhaps only the weekend or early next week.

On 14 January 2016 at 12:09, Jamie Bull <notifications@github.com> wrote:

> Thanks for taking a look at this. I've just spotted a couple of errors now
> I can see the diff - the *ScoreTuple.__repr_* and the accidentally
> deleted examples for the original GridSearchCV. I'll fix them once I've got
> your comments back as I'm hitting the hay now. Looking forward to your
> comments.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/6160#issuecomment-171490622>
> .
>

jnothman commented:
A first thought: if we're using a stability evaluation, this is more like CV and we need to run multiple fits per score evaluation. If not, then we're scoring based on the whole dataset (which may be appropriate, depending on application), so the notion of "refit" is redundant. Should this idea of averaging scores over multiple resamples be directly provided for, even to the extent of accepting a `cv` argument but only using the training set indices?

jnothman commented on sklearn/model_selection/_search.py:
```diff
@@ -4,40 +4,44 @@
 """
 from __future__ import print_function
 
-# Author: Alexandre Gramfort <alexandre.gramfort@inria.fr>,
```
Why did you move this?

jamiebull1 replied:
Not intentionally. It must be the default autopep8 settings in Eclipse. Is there a document with the expected autopep8 settings anywhere?

jamiebull1 replied:
Actually, it was the organise imports function in Eclipse that did this. I'll put these back as was with the next push and then disable that option.

amueller commented:
Can you give some examples?
I was thinking about this sometimes, but I thought the common use-case is already pretty easy:
```python
parm_iter = ParameterGrid(param_grid)
scores = [Est(**params).fit(X).score(X, y) for params in param_iter]
```
or something like that?

jamiebull1 commented:
That list pattern might be a good way of implementing some of the internals of this PR, but the intention is to allow users to have a similar API for finding the best parameters for transductive algorithms as they already have for algorithms which work with cross-validation. The similar, and so intuitive, API being the important part.

To give a concrete example (and the reason this idea came to me), I wanted to compare the various clustering algorithms in sklearn. To do this on a robust basis I obviously need to find the best set of hyperparameters for my dataset given my scoring criteria (in this case a combination of similar cluster size and cluster compactness). The first thing I thought of was to use the grid search cv class to search through the parameters. I'm sure that would be a common thought in someone with similar knowledge (basic to intermediate) of the package, so it seemed worth implementing. The fact that you thought about it too, even if only sometimes, suggests that it might be useful in practice.

jnothman commented:
@amueller, I think the advantages of this approach will be:
* support for different parameter search techniques (I assume GP-based optimisation pertains here too)
* potential support stability-based evaluation (averaging over training resamples) for which you sought in #1091
* similar interface to the `*SearchCV` classes that are familiar to many users (and hence easy to construct, fit, pickle, inspect...)

However, this approach also needs to be compared to #4301 wrt the extent of functionality.

jamiebull1 committed:
Revert moving of Authors and License section

jamiebull1 committed:
Revert moving of Authors and License section

jamiebull1 committed:
Fix test which was testing GridSearchCV not GridSearchTransductive

jamiebull1 committed:
Fix test which was testing GridSearchCV not GridSearchTransductive

jamiebull1 commented:
I've had a proper look at #4301 now and I don't see a huge amount of overlap. The metrics are clearly a useful addition and make having the `GridSearch` and `RandomSearch` framework more worthwhile. On the other hand, as currently implemented they only allow for the selection of a best value for `k` and not for other hyperparameters.

On your initial comments, @jnothman:

> A first thought: if we're using a stability evaluation, this is more like CV and we need to run multiple fits per score evaluation.

If I'm reading it right, this is what is calculated inside the `stability._one_stability_measure` in #4301 which returns a measure of cluster similarity. If this were exposed as the metric, rather than the current API of `stability.stability` which returns a best value for `k`, then it could be easily incorporated into a `scorer` for use in these transductive search classes.

>If not, then we're scoring based on the whole dataset (which may be appropriate, depending on application), so the notion of "refit" is redundant.

What I understood the refit to be doing is refitting the model using the best set of hyperparameters found during the search. But if I've misunderstood your point, let me know.

>Should this idea of averaging scores over multiple resamples be directly provided for, even to the extent of accepting a cv argument but only using the training set indices?

My thinking is that this would better placed as an argument to a `scorer` where this argument makes sense.

jamiebull1 commented:
@jnothman Is there any I should be doing here or just waiting to see what happens with #4301 ?

amueller commented:
I feel like #4301 is mostly unrelated: it implements scoring methods, while this PR implements the API for the search.

amueller commented:
Is transductive really the right word? Transductive is more like supervised learning when the test-set is known.

amueller commented on sklearn/model_selection/_search.py:
```diff
@@ -520,6 +548,456 @@ def inverse_transform(self, Xt):
+
+    Examples
+    --------
+    # TODO: Add examples here
```
todo?

amueller commented:
+1 on adding the functionality. I'm not sure about the naming, though. It should be obvious what it does from the name, to be discoverable. Also, it should not be confusing for people looking for the standard cross-validated search.

jamiebull1 commented:
@amueller Thanks for the review. Do you prefer the original names, `GridSearchCluster`, `RandomizedSeachCluster`, or would `GridSearchUnsupervised`, `RandomizedSeachUnsupervised` be better as less specific?

I agree that the #4301 scoring methods and this PR are complementary. They can probably be used together to make a good piece of documentation.

On the todo, I'll put together a doctest example to slot in there now I've received some positive feedback.

amueller commented:
I don't think ``cluster`` is necessarily good. And it doesn't really apply to all unsupervised methods. I don't really have any better ideas unfortunately. Maybe ``UnsupervisedGridSearch`` instead ``GridSearchUnsupervised`` would be a bit better? Or something that implies evaluating on the training set?

amueller commented:
@jnothman you proposed transductive, right? Do you have a reference for the use of the word referring to an algorithm used on a single dataset? I thought it meant that the test set is observed at training time.

amueller commented:
Ok even fictitous researchers can't agree really agree http://olivier.chapelle.cc/ssl-book/discussion.pdf. But I haven't heard of unsupervised algorithms being called transductive.

jnothman commented:
To my understanding (which I will admit is not because I deal with these things frequently), transductive learning has no intention to generalise beyond the particular dataset it is given: it learns a model of the instances it is given, some (but for obvious reasons not all) of which may be supervised. And of course, unsupervised algorithms can be inductive. What name do you give to unsupervised (including many uses of clustering, and t-SNE) and semi-supervised learning that is not intended to induce a model general to other instances, if not transduction? 

And sorry I've neglected in reviewing this. We're starting semester down here and together with conference deadlines, things are getting busy.

jamiebull1 committed:
Add doctest example for GridSearchTransductive

jamiebull1 committed:
Add doctest example for GridSearchTransductive

jamiebull1 commented:
On the naming, given the lack of an overarching term for the algorithms that could be used in this, could we perhaps name it for what it isn't rather than what it is? So `GridSearchNoCV` and `RandomizedSearchNoCV`. 

One more thing, have I been making these changes in the right place - `sklearn.model_selection._search.py`? Or should they be in `sklearn.grid_search.py`?

jamiebull1 commented:
Not sure how the coverage went down when I added a test. I assume other unrelated changes in the target branch?

amueller commented:
@jnothman I don't know of another word for it, but I haven't heard word ``transductive`` used to describe a clustering algorithm. Do you have any reference to that use of the word?

jnothman commented:
No, I don't!

rvraghav93 commented:
`UnsupervisedGridSearch` sounds good to me

GaelVaroquaux commented:
> UnsupervisedGridSearch sounds good to me

I like it: very explicit.

jamiebull1 commented:
Ok, so questions:

1. Should I go ahead with `UnsupervisedGridSearch`, etc. as it seems the most popular option?
2. Are my additions in the right module, or should they be in the `grid_search.py` module?
3. Should I be concerned about the failing coverage check?
4. Is there anything else I need to do?

jnothman commented:
Do go ahead with `UnsupervisedGridSearch`. I think it is more in line with people's understandings, even if IMO, there's not a strict requirement that this be unsupervised, nor that unsupervised models fail to generalise beyond their training and be evaluated on their generalisation capacity.

Don't be concerned about the coverage check.

And it appears you have not received any full reviews, so there'll probably be more to do. Sorry I've had a lot on my plate.

jamiebull1 committed:
Rename *SearchTransductive classes as Unsupervised*Search

jamiebull1 committed:
Rename *SearchTransductive classes as Unsupervised*Search

rvraghav93 commented:
Thanks for updating the PR. I'm taking a good look at this now...

rvraghav93 commented on sklearn/model_selection/_search.py:
```diff
@@ -14,30 +14,35 @@
 
 
 class ParameterGrid(object):
+
```
Why the extra line? (I feel this is the change of autopep8? ;) If so, I think, we generally don't recommend autopep8)

rvraghav93 commented on sklearn/model_selection/_search.py:
```diff
@@ -158,6 +163,7 @@ def __getitem__(self, ind):
 
 
 class ParameterSampler(object):
+
```
Same comment here

rvraghav93 commented on sklearn/model_selection/_search.py:
```diff
@@ -234,8 +241,8 @@ def __iter__(self):
-                    "than n_iter=%d." % (grid_size, self.n_iter)
-                    + " For exhaustive searches, use GridSearchCV.")
+                    "than n_iter=%d." % (grid_size, self.n_iter) +
+                    " For exhaustive searches, use GridSearchCV.")
```
Could you please avoid changing lines that are not part of your PR?
(That will help speed up the review of the actual content of your PR, helping you to merge it sooner into scikit-learn :) )

jamiebull1 replied:
Quite right, this is the dread hand of AutoPEP8. Apologies, I'll switch it
off in future.

On Tue, 1 Mar 2016, 14:54 Raghav R V, <notifications@github.com> wrote:

> In sklearn/model_selection/_search.py
> <https://github.com/scikit-learn/scikit-learn/pull/6160#discussion_r54568636>
> :
>
> > @@ -234,8 +241,8 @@ def __iter__(self):
> >              if grid_size < self.n_iter:
> >                  raise ValueError(
> >                      "The total space of parameters %d is smaller "
> > -                    "than n_iter=%d." % (grid_size, self.n_iter)
> > -                    + " For exhaustive searches, use GridSearchCV.")
> > +                    "than n_iter=%d." % (grid_size, self.n_iter) +
> > +                    " For exhaustive searches, use GridSearchCV.")
>
> Could you please avoid changing lines that are not part of your PR?
> (That will help speed up the review of the actual content of your PR,
> helping you to merge it sooner into scikit-learn :) )
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/6160/files#r54568636>.
>

rvraghav93 commented on sklearn/model_selection/_search.py:
```diff
@@ -360,9 +367,30 @@ def __repr__(self):
+        """Simple custom repr to summarize the main info"""
+        return "score: {0:.5f}, params: {1}".format(
+            self.score,
+            self.parameters)
```
@amueller This looks like a good way to store the `score` and `parameters`... WDYT?

rvraghav93 commented on sklearn/model_selection/_search.py:
```diff
@@ -360,9 +367,30 @@ def __repr__(self):
+
+class BaseSearch(six.with_metaclass(ABCMeta, BaseEstimator,
+                                    MetaEstimatorMixin)):
+
```
Same comment as the 1st one. Apologies for the nitpick ;(

rvraghav93 commented on sklearn/model_selection/_search.py:
```diff
@@ -520,6 +548,475 @@ def inverse_transform(self, Xt):
+        estimator = self.estimator
+
+        grid_scores = list()
+        out = Parallel(n_jobs=1
```
1 --> `self.n_jobs`?

rvraghav93 commented on sklearn/model_selection/_search.py:
```diff
@@ -520,6 +548,475 @@ def inverse_transform(self, Xt):
+            Target relative to X for classification or regression;
+            None for unsupervised learning.
+
+                        """
```
Alignment

rvraghav93 commented on sklearn/model_selection/_search.py:
```diff
@@ -520,6 +548,475 @@ def inverse_transform(self, Xt):
+
+class UnsupervisedRandomizedSearch(BaseSearch, UnsupervisedSearchMixin):
+
+    """Randomized search on hyper parameters.
```
... `"for unsupervised methods."`?

rvraghav93 commented on sklearn/model_selection/_search.py:
```diff
@@ -520,6 +548,475 @@ def inverse_transform(self, Xt):
+    See Also
+    --------
+    :class:`UnsupervisedGridSearch`:
+        Does exhaustive search over a grid of parameters.
```
... `"for supervised methods/algorithms"`?

rvraghav93 replied:
I'm removing this comment to reduce noise.

rvraghav93 commented on sklearn/model_selection/_search.py:
```diff
@@ -520,6 +548,475 @@ def inverse_transform(self, Xt):
+        Does exhaustive search over a grid of parameters.
+
+    :class:`ParameterSampler`:
+        A generator over parameter settins, constructed from
```
settings

rvraghav93 commented on sklearn/model_selection/tests/test_search.py:
```diff
@@ -1,69 +1,69 @@
+from sklearn.utils.testing import assert_raise_message
+from sklearn.utils.testing import assert_raises
+from sklearn.utils.testing import assert_warns
+from sklearn.utils.testing import ignore_warnings
```
Same comment as the previous one... Could you kindly avoid making changes not related to the PR. (This change is indeed welcome as it cleans up, but will blow up your diff and *may* distract reviewers.)

(Hope you don't mistake the aggressive nitpicking.)

rvraghav93 replied:
Could you revert this back to the original order? :)

rvraghav93 commented on sklearn/model_selection/tests/test_search.py:
```diff
@@ -168,6 +219,32 @@ def test_grid_search():
+
+    # Test exception handling on scoring
+    grid_search.scoring = 'sklearn'
+    assert_raises(ValueError, grid_search.fit, X)
```
+1 for neat unit tests. Thanks

jamiebull1 replied:
I can't take much credit for that. Just following the example of the great existing tests.

rvraghav93 commented on sklearn/model_selection/tests/test_search.py:
```diff
@@ -804,6 +1098,22 @@ def test_grid_search_failing_classifier_raise():
 
 
+def test_grid_search_failing_clusterer_raise():
+    # UnsupervisedGridSearch with on_error == 'raise' raises the error
```
with `error_score == 'raise'`

jamiebull1 replied:
Should I also fix this where it occurs in the CV tests?

rvraghav93 commented on sklearn/model_selection/tests/test_search.py:
```diff
@@ -336,6 +507,21 @@ def test_grid_search_one_grid_point():
     assert_array_equal(clf.dual_coef_, cv.best_estimator_.dual_coef_)
 
 
+@nottest
```
Why?

jamiebull1 replied:
Because of the TODO below. I'd forgotten about that, but have been given an answer on how to deal with it.

rvraghav93 replied:
Ah okay!

rvraghav93 commented on sklearn/model_selection/_search.py:
```diff
@@ -520,6 +548,475 @@ def inverse_transform(self, Xt):
+        # Find the best parameters by comparing on the score:
+        # note that `sorted` is deterministic in the way it breaks ties
+        best = sorted(grid_scores, key=lambda x: x.score,
+                      reverse=True)[0]
```
Why do we need to sort it? Can we not just return the one with the highest score? I understand this is how it was done previously for the regular `GridSearchCV`, but I am unable to figure out the reason. @amueller

rvraghav93 replied:
I mean is `max(grid_scores, key=lambda x: x.score)` not deterministic in breaking ties?

rvraghav93 commented:
@amueller @jnothman Probably a (very?) stupid suggestion, do you think it would be better to make `unsupervised` a parameter to `GridSearch` instead of having a separate class?

jamiebull1 committed:
Fix formatting and docstring issues identified in review
Remove blank lines before class docstrings
Reverted unnecessary format change
Formatting fix
Tweaks to docstrings

jamiebull1 committed:
Fix formatting and docstring issues identified in review
Remove blank lines before class docstrings
Reverted unnecessary format change
Formatting fix
Tweaks to docstrings

jamiebull1 committed:
Use n_jobs parameter in call to Parallel

jamiebull1 committed:
Use n_jobs parameter in call to Parallel

jamiebull1 committed:
Fix issues with tests identified in review
 
Test for labels normalized mutual info score to remove the issue where
differently-indexed labels caused a test to fail.
Test on blobs with greater separation to help KMeans identify clusters
Fix test name clash and incorrect docstrings in tests

jamiebull1 committed:
Fix issues with tests identified in review
 
Test for labels normalized mutual info score to remove the issue where
differently-indexed labels caused a test to fail.
Test on blobs with greater separation to help KMeans identify clusters
Fix test name clash and incorrect docstrings in tests

jamiebull1 commented:
Thanks for the review, @rvraghav93 . I hope I've addressed everything. I did consider whether it made more sense to add a parameter to `GridSearchCV`, or a special case for where k in k-folds = 1. In the end I thought that the name `GridSearchCV` was too explicitly about cross-validation, which this definitely isn't.

I also toyed with the idea of adding a warning suggesting the use of `UnsupervisedGridSearch` if the user tries to call `GridSearchCV` with k=1. If that sounds worthwhile then I can add that in.

jamiebull1 committed:
Remove unused import (nottest)

jamiebull1 committed:
Remove unused import (nottest)

jamiebull1 commented:
See #6474 for info on the failing check. It's unrelated to this PR.

jamiebull1 committed:
Revert to original import order (plus new imports)

jamiebull1 committed:
Revert to original import order (plus new imports)

jamiebull1 commented:
@rvraghav93  Is there anything else I can be doing on this over the weekend if I have some time?

rvraghav93 commented:
Thanks for addressing the comments! I'll take one more look at this and let you know in a while :)

rvraghav93 commented on sklearn/model_selection/__init__.py:
```diff
@@ -1,32 +1,34 @@
+from ._search import ParameterSampler
+from ._search import RandomizedSearchCV
+from ._search import UnsupervisedRandomizedSearch
+from ._search import fit_grid_point
```
revert the order to as it was before please? ;)

rvraghav93 replied:
Actually ignore this comment. This looks good and this is just in the `__init__.py`. Apologies for the noise!!

rvraghav93 replied:
Removing the comments to reduce noise in PR

rvraghav93 commented:
If you feel like it, you can -

1. Start modifying/adding docs for the new classes.
2. Whats new entry
3. One or Two didactic examples for `UnsupervisedGridSearch`, with some cool plots if needed! (Refer existing examples for gridsearch)

jamiebull1 committed:
Add whats new entry

jamiebull1 committed:
Add whats new entry

jamiebull1 committed:
Documentation changes

jamiebull1 committed:
Documentation changes

jamiebull1 commented:
Ok, great. I've put together a couple of notebooks using `UnsupervisedGridSearch` to find the best value for `n_clusters` and one using `UnsupervisedRandomSearch` for optimising hyperparameters for TSNE. Where do these need to go?

I've also added to the what's new and edited parts of the docstrings where I see it as necessary - there may be more changes required there that I've missed so I would appreciate a review of that aspect.

rvraghav93 commented:
It should be a python script with a description put inside the `examples` directory.

See [this directory for existing examples](https://github.com/scikit-learn/scikit-learn/tree/master/examples/model_selection).

jamiebull1 commented:
Thanks. Any idea why the tests started failing? I can't see any conflicts.

rvraghav93 commented:
Could you rebase upon the latest master once? There seems to be a merge conflict....

jnothman commented on sklearn/model_selection/_search.py:
```diff
@@ -520,6 +547,480 @@ def inverse_transform(self, Xt):
+    start_time = time.time()
+
+    try:
+        estimator.fit(X, y, **fit_params)
```
If the model is unsupervised, why should we provide `y`?

jnothman commented on sklearn/model_selection/_search.py:
```diff
@@ -520,6 +547,480 @@ def inverse_transform(self, Xt):
+
+    try:
+        estimator.fit(X, y, **fit_params)
+        score = scorer(estimator, X)
```
Just because the model is unsupervised (and non-inductive), why should it not have ground truth labels? Surely we need to support `scorer(estimator, X, y)` for rand score, v-measure, B-cubed, etc...

jamiebull1 replied:
Does this comment also answer the previous comment? At this point, `y` will either have been passed in or will be `None` and we should be able to pass either to the scorer. Or am I misunderstanding?

jnothman replied:
If unsupervised, `y` should not be provided to `fit`. But it may need to be provided to some scoring metrics. If semi-supervised, we're entering new territory.

jamiebull1 replied:
Thanks, that makes sense now. I'll make the changes and see if we need additional tests for scorers which use the known labels.

jnothman commented on sklearn/model_selection/_search.py:
```diff
@@ -520,6 +547,480 @@ def inverse_transform(self, Xt):
+              as in '2*n_jobs'
+
+    refit : boolean, default=True
+        Refit the best estimator with the entire dataset.
```
I don't think "entire" is appropriate here, because we're fitting with the entire dataset in all cases, are we not?

Indeed, it should be possible to retain (without refitting) the best model without any loss of efficiency, but that's not how things are currently implemented due to joblib not supporting an arbitrary reducer.

jamiebull1 committed:
Merge branch 'temp_for_rebase' into refactor_model_selection

jamiebull1 committed:
Add example scripts for unsupervised hyperparameter searches

rvraghav93 commented:
@jnothman What do you feel about having a single class for both unsupervised and supervised methods? (Not very explicit?)

jamiebull1 committed:
Allow passing y to unsupervised search when used in scorer

Don't pass y to estimator.fit in fit_and_score_unsupervised
Do pass y to scorer in fit_and_score_unsupervised

jnothman commented:
> What do you feel about having a single class for both unsupervised and supervised methods? 

I'm not quite going to answer this question yet.

The current state of this PR gives us no new functionality, as far as I can tell, but a slightly nicer interface (and, of course, examples which are great to have).  I think I can achieve roughly the current functionality of this PR with the helpers:
```python
class UnsupervisedCV(object):
    def split(self, X, y=None, labels=None):
        return [(np.arange(len(X)), np.arange(len(X)))]
    def get_n_splits(self, X, y=None, labels=None):
        return 1

def make_clusterer_truth_scorer(metric):
    return lambda estimator, X, y: metric(y, estimator.labels_)
```
then
```python
GridSearchCV(DBSCAN(), {'eps': [.1, .3, .5]}, cv=UnsupervisedCV(), scoring=make_clusterer_truth_scorer(v_measure_score)).fit(X, y)
```

So it's not already clear to me why we need to reduplicate the class structures for this use-case, except for the explicit naming. Would a not-so-special argument `cv='unsupervised'` suffice? Perhaps. (But it's basically a hack relying on the scorer being independent of `X`; it's a bit weird that the current proposal still has the scorer receiving `X`, yet of course that's more consistent for the user.)

Features/use-cases where it would not suffice, but currently not provided by this PR:

* if there were cases where calling `fit` then scorer was not possible, and `fit_predict`/`fit_transform` was needed directly.
* similarly, supporting a `fit_score` method as default scorer, but this doesn't currently exist on standard estimators
* refit not being performed redundantly (though this is not very exciting)
* the need to handle `y` with some labels missing for semi-supervision, but then for (some of) those labels to be available as ground truth
* potentially the need to only evaluate on a subset of the dataset that has ground truth (??)

jnothman commented:
Sorry for the perhaps overly harsh (but hopefully in a constructively incisive way) functional critique of this PR. I didn't really expect to silence all activity, and some thoughts on my issues would be very welcome from @jamiebull1, @amueller or @rvraghav93 among others.

jamiebull1 commented:
@jnothman It coincided with my supervisor finally coming back with constructive criticism on my dissertation. I'll try and make time to consider your points properly. I certainly don't have time to implement anything too involved any more though.

jnothman commented:
Oh dear. All the best for the dissertation. Don't worry about this. I'd
like to head feedback from the others in any case, about what they were
looking for in this enhancement.

On 31 March 2016 at 17:06, Jamie Bull <notifications@github.com> wrote:

> @jnothman <https://github.com/jnothman> It coincided with my supervisor
> finally coming back with constructive criticism on my dissertation. I'll
> try and make time to consider your points properly. I certainly don't have
> time to implement anything too involved any more though.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/6160#issuecomment-203770202>
>

vranjan90 commented:
Hii,
 Can anyone tell how to optimize clustering technique parameter using gidsearchcv or any other method using python???????

adrinjalali commented:
Hi @vranjan90 , this is a usage question and best asked on stackoverflow or other related forums.

adrinjalali commented:
Closing this as inactive. Leaving a comment on the issue for people who might want to continue this. There have been many issues/prs related to this, so it might be nice to add it at some point.
